src/lib.rs

```diff
@@ -88,42 +90,167 @@ impl InlinerBuilder {
+                for error in errors {
+                    writeln!(f, "* {}", error)?;
+                }
+                Ok(())
+            }
+        }
+    }
+}
+
+impl error::Error for Error {}
+
+impl Error {
+    /// Returns true if the error happened while reading or parsing the initial file.
+    pub fn is_initial(&self) -> bool {
+        match self {
+            Error::Initial(_) => true,
+            Error::Inline(_) => false,
+        }
+    }
+
+    /// Returns true if the error happened while inlining modules.
+    pub fn is_inline(&self) -> bool {
+        match self {
+            Error::Initial(_) => false,
+            Error::Inline(_) => true,
         }
     }
 }
 
-/// An error that was encountered while inlining modules
+/// The kind of error that was encountered for a particular file.
 #[derive(Debug)]
-pub enum Error {
-    /// The contents for one or more modules could not be found
-    NotFound(Vec<(String, SourceLocation)>),
-    #[doc(hidden)]
-    __NonExhaustive,
+#[non_exhaustive]
+pub enum ErrorKind {
+    /// An error happened while opening or reading the file.
+    Io(std::io::Error),
+
+    /// Errors happened while using `syn` to parse the file.
+    Parse(syn::Error),
+}
+
+impl From<std::io::Error> for ErrorKind {
+    fn from(err: std::io::Error) -> Self {
+        ErrorKind::Io(err)
+    }
+}
+
+impl From<syn::Error> for ErrorKind {
+    fn from(err: syn::Error) -> Self {
+        ErrorKind::Parse(err)
+    }
+}
+
+impl fmt::Display for ErrorKind {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            ErrorKind::Io(err) => write!(f, "IO error: {}", err),
+            ErrorKind::Parse(err) => write!(f, "parse error: {}", err),
+        }
+    }
+}
+
+/// An error that happened while attempting to inline a module.
+#[derive(Debug)]
+pub struct InlineError {
+    src_path: PathBuf,
+    module_name: String,
+    src_span: Span,
+    path: PathBuf,
+    kind: ErrorKind,
+}
+
+impl InlineError {
+    pub(crate) fn new(
+        src_path: impl Into<PathBuf>,
+        item_mod: &ItemMod,
+        path: impl Into<PathBuf>,
+        kind: ErrorKind,
+    ) -> Self {
+        Self {
+            src_path: src_path.into(),
+            module_name: item_mod.ident.to_string(),
+            src_span: item_mod.span(),
+            path: path.into(),
+            kind,
+        }
+    }
+
+    /// Returns the source path where the error originated.
+    ///
+    /// The file at this path parsed correctly, but it caused the file at `self.path()` to be read.
+    pub fn src_path(&self) -> &Path {
+        &self.src_path
+    }
+
+    /// Returns the name of the module that was attempted to be inlined.
+    pub fn module_name(&self) -> &str {
+        &self.module_name
+    }
+
+    /// Returns the `Span` (including line and column information) in the source path that caused
+    /// `self.path()` to be included.
+    pub fn src_span(&self) -> proc_macro2::Span {
```

Comment:
```suggestion
    pub fn src_span(&self) -> Span {
```

If you were deliberately fully qualifying the path here for rustdoc's benefit, feel free to ignore this suggestion. Otherwise, since we have the type locally imported it seems cleaner not to repeat the crate name.